### ISSUE TITLE
[Question]JSDOM alternatives in Playwright

### ISSUE THREAD

**Issue by hassan-aziz** [microsoft/playwright#15507]
Currently the front end application is using JEST as a test runner and JSDOM to emulate the browser. With this approach it is possible to call "window" object and use something from the application code in the test case eg "window.sessionStorage.clear()". After I changed the test cases from JEST to Playwright runner, I can no longer run the tests due to error "ReferenceError: window is not defined". This is because "window" object doesn't exist on node.js. Is there any alternative in playwright to run the tests in different environment than node.js so that it is possible to use "window"?

One of the existing solution is to use "await page.evaluate(window.sessionStorage.clear())",  but it only helps in test files, if the application code is still required (calling a function from application code base to set something in window) then it fails again because window object don't exists.

In short I would like to replace the JEST runner completely with Playwright runner

**Comment by yury-s** [comment]
> Is there any alternative in playwright to run the tests in different environment than node.js so that it is possible to use "window"?

The test code always runs in node.js while the web application is loaded into a browser in playwright. Window object is only available to the code that runs inside the browser.

> One of the existing solution is to use "await page.evaluate(window.sessionStorage.clear())", but it only helps in test files, if the application code is still required (calling a function from application code base to set something in window) then it fails again because window object don't exists.

What you describe is the right way to interact with the web application code in playwright. The test code runs in the node process while web application is in the browser, so you use playwright objects (page, locators) to interact with the application. Similar to clearing the storage you can call arbitrary functions inside the web app: 

```js
await page.evaluate(() => window.yourAppGlobalFunction());
```

You can use  [`context.exposeBinding`](https://playwright.dev/docs/next/api/class-browsercontext#browser-context-expose-binding) to insert some functions into the web app code that would trigger function calls on the test side.

Perhaps you can share a specific code snippet that doesn't work for you and we could help you improve it?

**Comment by yury-s** [comment]
Closing as part of the triage process since it seemed stale. Please create a new issue with a detailed reproducible or feature request if you still face issues.
